### DISCUSSION TITLE
[Proposal]: Why aren't both String.Substring()'s arguments indexes?

### DISCUSSION THREAD

**Discussion by yaroshchuk8** [dotnet/csharplang#6522] [Language Ideas] [answered]
I've spent several hours thinking about what's wrong with my code and eventually realized that I misunderstood the way String.Substring() works. I thought it takes 2 indexes and returns a new string according to them, but there's some length argument that doesn't make any sense to me. 

So, my proposal is to change this method: it has to take two indexes (firstIndex, lastIndex) as arguments and return a new string.

There's a simple function I wrote to solve my problem:
```
string Sub(string s, int firstIndex, int lastIndex)
{
    char[] temp = new char[lastIndex - firstIndex + 1];
    
    for (int counter = 0, i = firstIndex; i <= lastIndex; i++, counter++)
    {
        temp[counter] = s[i];
    }

    return new string(temp);
}
```
This version is flexible because it includes the same argument implicitly (to transform [lastIndex] into [length] write [firstIndex+ whatever your length is]).

P.S. I realize that I can be wrong.

**Comment by HaloFour** [comment, marked as answer]
The [runtime](https://github.com/dotnet/runtime) repository would be the appropriate location for suggesting changes and additions to the API.

> So, my proposal is to change this method

Changing the existing method would be a non-starter.  The method has been around for 20 years and is used by an _enormous_ amount of existing code which would now be broken.  At best a new method could be suggested.

**Reply by yaroshchuk8** [reply]
Got it.

**Comment by CyrusNajmabadi** [comment]
String is also slicable.  So you can just do `myString[index1..index2]` and that will work properly.

**Reply by yaroshchuk8** [reply]
Thank you.
